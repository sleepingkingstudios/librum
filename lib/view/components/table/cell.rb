# frozen_string_literal: true

require 'view/components/table'

module View::Components
  # Component for rendering individual data cells.
  class Table::Cell < View::Components::DataField
    # @param column [View::Components::DataField::FieldDefinition] the column
    #   used to render the cell.
    # @param data [Hash{String=>Object}] the data used to render the cell.
    def initialize(column:, data:, **)
      super(data: data, field: column)
    end
  end
end
