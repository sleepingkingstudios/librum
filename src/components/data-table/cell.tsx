import * as React from 'react';
import { kebabCase } from 'lodash';

import {
  isComponent,
  joinClassNames,
} from '@utils/react-utils';
import type {
  ColumnValueFunction,
  DataTableColumn,
} from './types';
import {
  alignmentClass,
  labelFor,
} from './utils';

type CellProps = {
  column: DataTableColumn,
  data: Record<string, unknown>,
};

const defaultFor = ({
  column,
  data,
}: CellProps): JSX.Element | string => {
  if (!('default' in column)) { return ' '; }

  if (typeof column.default === 'string') { return column.default; }

  if (typeof column.default === 'function') {
    if (isComponent(column.default)) {
      const DefaultComponent = column.default as React.ComponentType<CellProps>;

      return (<DefaultComponent column={column} data={data} />);
    }

    const fn = column.default as ColumnValueFunction;

    return fn({ data });
  }

  return ' ';
};

const isEmpty = ({ value }: { value: unknown }): boolean => {
  if (value === null) { return true; }

  if (value === undefined) { return true; }

  if (typeof value === 'string' && value.length === 0) { return true; }

  return false;
};

const valueFor = ({
  column,
  data,
}: CellProps): JSX.Element | string => {
  if (typeof column.value === 'function') {
    if (isComponent(column.value)) {
      const ValueComponent = column.value as React.ComponentType<CellProps>;

      return (<ValueComponent column={column} data={data} />);
    }

    const fn = column.value as ColumnValueFunction;

    return fn({ data });
  }

  if (column.name in data) {
    const value = data[column.name];

    if (isEmpty({ value })) { return defaultFor({ column, data }); }

    return value.toString();
  }

  return defaultFor({ column, data });
};

export const DataTableCell = ({
  column,
  data,
}: CellProps): JSX.Element => {
  const joinedClassName = joinClassNames(
    `data-table-cell-${kebabCase(column.name)}`,
    alignmentClass(column),
    column.numeric ? 'lining-nums tabular-nums' : null,
  );
  const label = labelFor({ column, force: true });
  const value = valueFor({ column, data });

  return (
    <td className={joinedClassName} data-label={label}>{ value }</td>
  );
};
