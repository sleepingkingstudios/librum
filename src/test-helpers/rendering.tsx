import * as React from 'react';
import { Provider as ReduxProvider } from 'react-redux';
import {
  MemoryRouter,
  Routes,
  Route
} from "react-router-dom";
import {
  RenderOptions,
  RenderResult,
  render,
} from '@testing-library/react';

import {
  createStore,
  storeType,
} from './store';
import {
  ThemeContext,
  defaultTheme,
} from '@themes/context';
import type { Theme } from '@themes';

interface customRenderOptions extends RenderOptions {
  initialEntries?: Array<string>;
  router?: true | withRouterOptions;
  store?: (storeType | true);
  theme?: (Theme | true);
}

interface withRouterOptions {
  initialEntries?: Array<string>;
  path?: string;
}

export const withRouter = (
  component: React.ReactElement,
  {
    initialEntries = ['/'],
    path,
  }: withRouterOptions,
): React.ReactElement => {
  let wrapped = component;

  if (path) {
    wrapped = (
      <Routes>
        <Route path={path} element={wrapped} />
      </Routes>
    );
  }

  return (
    <MemoryRouter initialEntries={initialEntries}>
      { wrapped }
    </MemoryRouter>
  );
};

export const withStore = (
  component: React.ReactElement,
  { store }: { store: (storeType | true) }
): React.ReactElement => {
  let configuredStore: storeType;

  if (store === true) {
    configuredStore = createStore().store;
  } else {
    configuredStore = store;
  }

  return (
    <ReduxProvider store={configuredStore}>
      { component }
    </ReduxProvider>
  );
};

export const withTheme = (
  component: React.ReactElement,
  { theme }: { theme: (Theme | true) }
): React.ReactElement => {
  let configuredTheme: Theme;

  if (theme === true) {
    configuredTheme = defaultTheme;
  } else {
    configuredTheme = theme;
  }

  return (
    <ThemeContext.Provider value={configuredTheme}>
      { component }
    </ThemeContext.Provider>
  );
};

const customRender = (
  component: React.ReactElement,
  options?: customRenderOptions,
  strictMode = true,
): RenderResult => {
  let wrapped: React.ReactElement = component;

  if (strictMode) {
    wrapped = (<React.StrictMode>{ wrapped }</React.StrictMode>);
  }

  if (!options) { return render(wrapped); }

  if ('router' in options && options.router) {
    if (options.router === true) {
      wrapped = withRouter(wrapped, { initialEntries: options.initialEntries });
    } else {
      wrapped = withRouter(
        wrapped,
        {
          initialEntries: options.initialEntries,
          ...options.router,
        },
      );
    }
  }

  if ('store' in options) {
    wrapped = withStore(wrapped, { store: options.store });
  }

  if ('theme' in options) {
    wrapped = withTheme(wrapped, { theme: options.theme });
  }

  return render(wrapped, options);
};

export { customRender as render };
