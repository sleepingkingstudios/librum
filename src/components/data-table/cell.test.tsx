import * as React from 'react';

import '@testing-library/jest-dom';
import { render } from '@test-helpers/rendering';
import type { DataTableColumn } from './types';

import { DataTableCell } from './cell';

type SpaceMission = {
  missionName?: string,
  ordinal?: number,
  program?: string,
};

const CellWrapper = ({
  children,
}: {
  children: React.ReactNode,
}): JSX.Element => (<table><tbody><tr>{ children }</tr></tbody></table>);

describe('<DataTableCell />', () => {
  describe('with a basic column', () => {
    const column: DataTableColumn = { name: 'missionName' };
    const data = { missionName: 'Hephaestus VI' };

    it('should display the data', () => {
      const { getByRole } = render(
        <DataTableCell column={column} data={data} />,
        { wrapper: CellWrapper },
      );
      const cell = getByRole('cell');

      expect(cell).toBeVisible();
      expect(cell.textContent).toBe('Hephaestus VI');
    });

    it('should match the snapshot', () => {
      const { asFragment } = render(
        <DataTableCell column={column} data={data} />,
        { wrapper: CellWrapper },
      );

      expect(asFragment()).toMatchSnapshot();
    });

    describe('when the data does not have a matching property', () => {
      const nonMatching = {};

      it('should display an empty cell', () => {
        const { getByRole } = render(
          <DataTableCell column={column} data={nonMatching} />,
          { wrapper: CellWrapper },
        );
        const cell = getByRole('cell');

        expect(cell).toBeVisible();
        expect(cell.textContent).toBe(' ');
      });

      it('should match the snapshot', () => {
        const { asFragment } = render(
          <DataTableCell column={column} data={nonMatching} />,
          { wrapper: CellWrapper },
        );

        expect(asFragment()).toMatchSnapshot();
      });
    });

    describe('when the data has a null property', () => {
      const empty: SpaceMission = { missionName: null };

      it('should display an empty cell', () => {
        const { getByRole } = render(
          <DataTableCell column={column} data={empty} />,
          { wrapper: CellWrapper },
        );
        const cell = getByRole('cell');

        expect(cell).toBeVisible();
        expect(cell.textContent).toBe(' ');
      });

      it('should match the snapshot', () => {
        const { asFragment } = render(
          <DataTableCell column={column} data={empty} />,
          { wrapper: CellWrapper },
        );

        expect(asFragment()).toMatchSnapshot();
      });
    });

    describe('when the data has an empty property', () => {
      const empty = { missionName: '' };

      it('should display an empty cell', () => {
        const { getByRole } = render(
          <DataTableCell column={column} data={empty} />,
          { wrapper: CellWrapper },
        );
        const cell = getByRole('cell');

        expect(cell).toBeVisible();
        expect(cell.textContent).toBe(' ');
      });

      it('should match the snapshot', () => {
        const { asFragment } = render(
          <DataTableCell column={column} data={empty} />,
          { wrapper: CellWrapper },
        );

        expect(asFragment()).toMatchSnapshot();
      });
    });

    describe('when the data has an undefined property', () => {
      const empty: SpaceMission = { missionName: undefined };

      it('should display an empty cell', () => {
        const { getByRole } = render(
          <DataTableCell column={column} data={empty} />,
          { wrapper: CellWrapper },
        );
        const cell = getByRole('cell');

        expect(cell).toBeVisible();
        expect(cell.textContent).toBe(' ');
      });

      it('should match the snapshot', () => {
        const { asFragment } = render(
          <DataTableCell column={column} data={empty} />,
          { wrapper: CellWrapper },
        );

        expect(asFragment()).toMatchSnapshot();
      });
    });
  });

  describe('with a column with align: value', () => {
    const column: DataTableColumn = {
      align: 'center',
      name: 'missionName',
    };
    const data = { missionName: 'Hephaestus VI' };

    it('should match the snapshot', () => {
      const { asFragment } = render(
        <DataTableCell column={column} data={data} />,
        { wrapper: CellWrapper },
      );

      expect(asFragment()).toMatchSnapshot();
    });
  });

  describe('with a column with default: component', () => {
    const Default = ({ data }: { data: SpaceMission }): JSX.Element => (
      <span className="data-table-cell-default">
        { `${data.program} ?` }
      </span>
    );
    const column: DataTableColumn = {
      default: Default,
      name: 'missionName',
    };
    const data = {
      missionName: 'Hephaestus VI',
      program: 'Hephaestus',
    };

    it('should display the data', () => {
      const { getByRole } = render(
        <DataTableCell column={column} data={data} />,
        { wrapper: CellWrapper },
      );
      const cell = getByRole('cell');

      expect(cell).toBeVisible();
      expect(cell.textContent).toBe('Hephaestus VI');
    });

    it('should match the snapshot', () => {
      const { asFragment } = render(
        <DataTableCell column={column} data={data} />,
        { wrapper: CellWrapper },
      );

      expect(asFragment()).toMatchSnapshot();
    });

    describe('when the data does not have a matching property', () => {
      const nonMatching = { program: 'Hephaestus' };

      it('should display the default value', () => {
        const { getByRole } = render(
          <DataTableCell column={column} data={nonMatching} />,
          { wrapper: CellWrapper },
        );
        const cell = getByRole('cell');

        expect(cell).toBeVisible();
        expect(cell.textContent).toBe('Hephaestus ?');
      });

      it('should match the snapshot', () => {
        const { asFragment } = render(
          <DataTableCell column={column} data={nonMatching} />,
          { wrapper: CellWrapper },
        );

        expect(asFragment()).toMatchSnapshot();
      });
    });

    describe('when the data has a null property', () => {
      const empty: SpaceMission = { missionName: null, program: 'Hephaestus' };

      it('should display the default value', () => {
        const { getByRole } = render(
          <DataTableCell column={column} data={empty} />,
          { wrapper: CellWrapper },
        );
        const cell = getByRole('cell');

        expect(cell).toBeVisible();
        expect(cell.textContent).toBe('Hephaestus ?');
      });

      it('should match the snapshot', () => {
        const { asFragment } = render(
          <DataTableCell column={column} data={empty} />,
          { wrapper: CellWrapper },
        );

        expect(asFragment()).toMatchSnapshot();
      });
    });

    describe('when the data has an empty property', () => {
      const empty = { missionName: '', program: 'Hephaestus' };

      it('should display the default value', () => {
        const { getByRole } = render(
          <DataTableCell column={column} data={empty} />,
          { wrapper: CellWrapper },
        );
        const cell = getByRole('cell');

        expect(cell).toBeVisible();
        expect(cell.textContent).toBe('Hephaestus ?');
      });

      it('should match the snapshot', () => {
        const { asFragment } = render(
          <DataTableCell column={column} data={empty} />,
          { wrapper: CellWrapper },
        );

        expect(asFragment()).toMatchSnapshot();
      });
    });

    describe('when the data has an undefined property', () => {
      const empty: SpaceMission = {
        missionName: undefined,
        program: 'Hephaestus',
      };

      it('should display the default value', () => {
        const { getByRole } = render(
          <DataTableCell column={column} data={empty} />,
          { wrapper: CellWrapper },
        );
        const cell = getByRole('cell');

        expect(cell).toBeVisible();
        expect(cell.textContent).toBe('Hephaestus ?');
      });

      it('should match the snapshot', () => {
        const { asFragment } = render(
          <DataTableCell column={column} data={empty} />,
          { wrapper: CellWrapper },
        );

        expect(asFragment()).toMatchSnapshot();
      });
    });
  });

  describe('with a column with default: function', () => {
    const fn = ({
      data,
    }: {
      data: SpaceMission,
    }): string => `${data.program} ?`;
    const column: DataTableColumn = {
      default: fn,
      name: 'missionName',
    };
    const data = {
      missionName: 'Hephaestus VI',
      program: 'Hephaestus',
    };

    it('should display the data', () => {
      const { getByRole } = render(
        <DataTableCell column={column} data={data} />,
        { wrapper: CellWrapper },
      );
      const cell = getByRole('cell');

      expect(cell).toBeVisible();
      expect(cell.textContent).toBe('Hephaestus VI');
    });

    it('should match the snapshot', () => {
      const { asFragment } = render(
        <DataTableCell column={column} data={data} />,
        { wrapper: CellWrapper },
      );

      expect(asFragment()).toMatchSnapshot();
    });

    describe('when the data does not have a matching property', () => {
      const nonMatching = { program: 'Hephaestus' };

      it('should display the default value', () => {
        const { getByRole } = render(
          <DataTableCell column={column} data={nonMatching} />,
          { wrapper: CellWrapper },
        );
        const cell = getByRole('cell');

        expect(cell).toBeVisible();
        expect(cell.textContent).toBe('Hephaestus ?');
      });

      it('should match the snapshot', () => {
        const { asFragment } = render(
          <DataTableCell column={column} data={nonMatching} />,
          { wrapper: CellWrapper },
        );

        expect(asFragment()).toMatchSnapshot();
      });
    });

    describe('when the data has a null property', () => {
      const empty: SpaceMission = { missionName: null, program: 'Hephaestus' };

      it('should display the default value', () => {
        const { getByRole } = render(
          <DataTableCell column={column} data={empty} />,
          { wrapper: CellWrapper },
        );
        const cell = getByRole('cell');

        expect(cell).toBeVisible();
        expect(cell.textContent).toBe('Hephaestus ?');
      });

      it('should match the snapshot', () => {
        const { asFragment } = render(
          <DataTableCell column={column} data={empty} />,
          { wrapper: CellWrapper },
        );

        expect(asFragment()).toMatchSnapshot();
      });
    });

    describe('when the data has an empty property', () => {
      const empty = { missionName: '', program: 'Hephaestus' };

      it('should display the default value', () => {
        const { getByRole } = render(
          <DataTableCell column={column} data={empty} />,
          { wrapper: CellWrapper },
        );
        const cell = getByRole('cell');

        expect(cell).toBeVisible();
        expect(cell.textContent).toBe('Hephaestus ?');
      });

      it('should match the snapshot', () => {
        const { asFragment } = render(
          <DataTableCell column={column} data={empty} />,
          { wrapper: CellWrapper },
        );

        expect(asFragment()).toMatchSnapshot();
      });
    });

    describe('when the data has an undefined property', () => {
      const empty: SpaceMission = {
        missionName: undefined,
        program: 'Hephaestus',
      };

      it('should display the default value', () => {
        const { getByRole } = render(
          <DataTableCell column={column} data={empty} />,
          { wrapper: CellWrapper },
        );
        const cell = getByRole('cell');

        expect(cell).toBeVisible();
        expect(cell.textContent).toBe('Hephaestus ?');
      });

      it('should match the snapshot', () => {
        const { asFragment } = render(
          <DataTableCell column={column} data={empty} />,
          { wrapper: CellWrapper },
        );

        expect(asFragment()).toMatchSnapshot();
      });
    });
  });

  describe('with a column with default: string', () => {
    const column: DataTableColumn = {
      default: 'N/A',
      name: 'missionName',
    };
    const data = { missionName: 'Hephaestus VI' };

    it('should display the data', () => {
      const { getByRole } = render(
        <DataTableCell column={column} data={data} />,
        { wrapper: CellWrapper },
      );
      const cell = getByRole('cell');

      expect(cell).toBeVisible();
      expect(cell.textContent).toBe('Hephaestus VI');
    });

    it('should match the snapshot', () => {
      const { asFragment } = render(
        <DataTableCell column={column} data={data} />,
        { wrapper: CellWrapper },
      );

      expect(asFragment()).toMatchSnapshot();
    });

    describe('when the data does not have a matching property', () => {
      const nonMatching = {};

      it('should display the default value', () => {
        const { getByRole } = render(
          <DataTableCell column={column} data={nonMatching} />,
          { wrapper: CellWrapper },
        );
        const cell = getByRole('cell');

        expect(cell).toBeVisible();
        expect(cell.textContent).toBe('N/A');
      });

      it('should match the snapshot', () => {
        const { asFragment } = render(
          <DataTableCell column={column} data={nonMatching} />,
          { wrapper: CellWrapper },
        );

        expect(asFragment()).toMatchSnapshot();
      });
    });

    describe('when the data has a null property', () => {
      const empty: SpaceMission = { missionName: null };

      it('should display the default value', () => {
        const { getByRole } = render(
          <DataTableCell column={column} data={empty} />,
          { wrapper: CellWrapper },
        );
        const cell = getByRole('cell');

        expect(cell).toBeVisible();
        expect(cell.textContent).toBe('N/A');
      });

      it('should match the snapshot', () => {
        const { asFragment } = render(
          <DataTableCell column={column} data={empty} />,
          { wrapper: CellWrapper },
        );

        expect(asFragment()).toMatchSnapshot();
      });
    });

    describe('when the data has an empty property', () => {
      const empty = { missionName: '' };

      it('should display the default value', () => {
        const { getByRole } = render(
          <DataTableCell column={column} data={empty} />,
          { wrapper: CellWrapper },
        );
        const cell = getByRole('cell');

        expect(cell).toBeVisible();
        expect(cell.textContent).toBe('N/A');
      });

      it('should match the snapshot', () => {
        const { asFragment } = render(
          <DataTableCell column={column} data={empty} />,
          { wrapper: CellWrapper },
        );

        expect(asFragment()).toMatchSnapshot();
      });
    });

    describe('when the data has an undefined property', () => {
      const empty: SpaceMission = { missionName: undefined };

      it('should display the default value', () => {
        const { getByRole } = render(
          <DataTableCell column={column} data={empty} />,
          { wrapper: CellWrapper },
        );
        const cell = getByRole('cell');

        expect(cell).toBeVisible();
        expect(cell.textContent).toBe('N/A');
      });

      it('should match the snapshot', () => {
        const { asFragment } = render(
          <DataTableCell column={column} data={empty} />,
          { wrapper: CellWrapper },
        );

        expect(asFragment()).toMatchSnapshot();
      });
    });
  });

  describe('with a column with numeric: true', () => {
    const column: DataTableColumn = {
      name: 'deltaV',
      numeric: true,
    };
    const data = { deltaV: '1,000 km/s' };

    it('should match the snapshot', () => {
      const { asFragment } = render(
        <DataTableCell column={column} data={data} />,
        { wrapper: CellWrapper },
      );

      expect(asFragment()).toMatchSnapshot();
    });
  });

  describe('with a column with value: component', () => {
    const MissionName = ({ data }: { data: SpaceMission }): JSX.Element => (
      <span className="data-table-cell-mission-name">
        { `${data.missionName} ${data.ordinal}` }
      </span>
    );
    const column: DataTableColumn = {
      name: 'missionName',
      value: MissionName,
    };
    const data = {
      missionName: 'Hephaestus',
      ordinal: 'VI',
    };

    it('should display the data', () => {
      const { getByRole } = render(
        <DataTableCell column={column} data={data} />,
        { wrapper: CellWrapper },
      );
      const cell = getByRole('cell');

      expect(cell).toBeVisible();
      expect(cell.textContent).toBe('Hephaestus VI');
    });

    it('should match the snapshot', () => {
      const { asFragment } = render(
        <DataTableCell column={column} data={data} />,
        { wrapper: CellWrapper },
      );

      expect(asFragment()).toMatchSnapshot();
    });
  });

  describe('with a column with value: function', () => {
    const fn = ({
      data,
    }: {
      data: SpaceMission,
    }): string => `${data.missionName} ${data.ordinal}`;
    const column: DataTableColumn = {
      name: 'missionName',
      value: fn,
    };
    const data = {
      missionName: 'Hephaestus',
      ordinal: 'VI',
    };

    it('should display the data', () => {
      const { getByRole } = render(
        <DataTableCell column={column} data={data} />,
        { wrapper: CellWrapper },
      );
      const cell = getByRole('cell');

      expect(cell).toBeVisible();
      expect(cell.textContent).toBe('Hephaestus VI');
    });

    it('should match the snapshot', () => {
      const { asFragment } = render(
        <DataTableCell column={column} data={data} />,
        { wrapper: CellWrapper },
      );

      expect(asFragment()).toMatchSnapshot();
    });
  });
});
