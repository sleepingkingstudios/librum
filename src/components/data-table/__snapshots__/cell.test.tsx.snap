// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DataTableCell /> with a basic column should match the snapshot 1`] = `
<DocumentFragment>
  <table>
    <tbody>
      <tr>
        <td
          class="data-table-cell-mission-name"
          data-label="Mission Name"
        >
          Hephaestus VI
        </td>
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;

exports[`<DataTableCell /> with a basic column when the data does not have a matching property should match the snapshot 1`] = `
<DocumentFragment>
  <table>
    <tbody>
      <tr>
        <td
          class="data-table-cell-mission-name"
          data-label="Mission Name"
        >
           
        </td>
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;

exports[`<DataTableCell /> with a basic column when the data has a null property should match the snapshot 1`] = `
<DocumentFragment>
  <table>
    <tbody>
      <tr>
        <td
          class="data-table-cell-mission-name"
          data-label="Mission Name"
        >
           
        </td>
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;

exports[`<DataTableCell /> with a basic column when the data has an empty property should match the snapshot 1`] = `
<DocumentFragment>
  <table>
    <tbody>
      <tr>
        <td
          class="data-table-cell-mission-name"
          data-label="Mission Name"
        >
           
        </td>
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;

exports[`<DataTableCell /> with a basic column when the data has an undefined property should match the snapshot 1`] = `
<DocumentFragment>
  <table>
    <tbody>
      <tr>
        <td
          class="data-table-cell-mission-name"
          data-label="Mission Name"
        >
           
        </td>
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;

exports[`<DataTableCell /> with a column with align: value should match the snapshot 1`] = `
<DocumentFragment>
  <table>
    <tbody>
      <tr>
        <td
          class="data-table-cell-mission-name text-center"
          data-label="Mission Name"
        >
          Hephaestus VI
        </td>
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;

exports[`<DataTableCell /> with a column with default: component should match the snapshot 1`] = `
<DocumentFragment>
  <table>
    <tbody>
      <tr>
        <td
          class="data-table-cell-mission-name"
          data-label="Mission Name"
        >
          Hephaestus VI
        </td>
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;

exports[`<DataTableCell /> with a column with default: component when the data does not have a matching property should match the snapshot 1`] = `
<DocumentFragment>
  <table>
    <tbody>
      <tr>
        <td
          class="data-table-cell-mission-name"
          data-label="Mission Name"
        >
          <span
            class="data-table-cell-default"
          >
            Hephaestus ?
          </span>
        </td>
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;

exports[`<DataTableCell /> with a column with default: component when the data has a null property should match the snapshot 1`] = `
<DocumentFragment>
  <table>
    <tbody>
      <tr>
        <td
          class="data-table-cell-mission-name"
          data-label="Mission Name"
        >
          <span
            class="data-table-cell-default"
          >
            Hephaestus ?
          </span>
        </td>
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;

exports[`<DataTableCell /> with a column with default: component when the data has an empty property should match the snapshot 1`] = `
<DocumentFragment>
  <table>
    <tbody>
      <tr>
        <td
          class="data-table-cell-mission-name"
          data-label="Mission Name"
        >
          <span
            class="data-table-cell-default"
          >
            Hephaestus ?
          </span>
        </td>
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;

exports[`<DataTableCell /> with a column with default: component when the data has an undefined property should match the snapshot 1`] = `
<DocumentFragment>
  <table>
    <tbody>
      <tr>
        <td
          class="data-table-cell-mission-name"
          data-label="Mission Name"
        >
          <span
            class="data-table-cell-default"
          >
            Hephaestus ?
          </span>
        </td>
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;

exports[`<DataTableCell /> with a column with default: function should match the snapshot 1`] = `
<DocumentFragment>
  <table>
    <tbody>
      <tr>
        <td
          class="data-table-cell-mission-name"
          data-label="Mission Name"
        >
          Hephaestus VI
        </td>
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;

exports[`<DataTableCell /> with a column with default: function when the data does not have a matching property should match the snapshot 1`] = `
<DocumentFragment>
  <table>
    <tbody>
      <tr>
        <td
          class="data-table-cell-mission-name"
          data-label="Mission Name"
        >
          Hephaestus ?
        </td>
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;

exports[`<DataTableCell /> with a column with default: function when the data has a null property should match the snapshot 1`] = `
<DocumentFragment>
  <table>
    <tbody>
      <tr>
        <td
          class="data-table-cell-mission-name"
          data-label="Mission Name"
        >
          Hephaestus ?
        </td>
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;

exports[`<DataTableCell /> with a column with default: function when the data has an empty property should match the snapshot 1`] = `
<DocumentFragment>
  <table>
    <tbody>
      <tr>
        <td
          class="data-table-cell-mission-name"
          data-label="Mission Name"
        >
          Hephaestus ?
        </td>
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;

exports[`<DataTableCell /> with a column with default: function when the data has an undefined property should match the snapshot 1`] = `
<DocumentFragment>
  <table>
    <tbody>
      <tr>
        <td
          class="data-table-cell-mission-name"
          data-label="Mission Name"
        >
          Hephaestus ?
        </td>
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;

exports[`<DataTableCell /> with a column with default: string should match the snapshot 1`] = `
<DocumentFragment>
  <table>
    <tbody>
      <tr>
        <td
          class="data-table-cell-mission-name"
          data-label="Mission Name"
        >
          Hephaestus VI
        </td>
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;

exports[`<DataTableCell /> with a column with default: string when the data does not have a matching property should match the snapshot 1`] = `
<DocumentFragment>
  <table>
    <tbody>
      <tr>
        <td
          class="data-table-cell-mission-name"
          data-label="Mission Name"
        >
          N/A
        </td>
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;

exports[`<DataTableCell /> with a column with default: string when the data has a null property should match the snapshot 1`] = `
<DocumentFragment>
  <table>
    <tbody>
      <tr>
        <td
          class="data-table-cell-mission-name"
          data-label="Mission Name"
        >
          N/A
        </td>
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;

exports[`<DataTableCell /> with a column with default: string when the data has an empty property should match the snapshot 1`] = `
<DocumentFragment>
  <table>
    <tbody>
      <tr>
        <td
          class="data-table-cell-mission-name"
          data-label="Mission Name"
        >
          N/A
        </td>
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;

exports[`<DataTableCell /> with a column with default: string when the data has an undefined property should match the snapshot 1`] = `
<DocumentFragment>
  <table>
    <tbody>
      <tr>
        <td
          class="data-table-cell-mission-name"
          data-label="Mission Name"
        >
          N/A
        </td>
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;

exports[`<DataTableCell /> with a column with numeric: true should match the snapshot 1`] = `
<DocumentFragment>
  <table>
    <tbody>
      <tr>
        <td
          class="data-table-cell-delta-v text-right lining-nums tabular-nums"
          data-label="Delta V"
        >
          1,000 km/s
        </td>
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;

exports[`<DataTableCell /> with a column with value: component should match the snapshot 1`] = `
<DocumentFragment>
  <table>
    <tbody>
      <tr>
        <td
          class="data-table-cell-mission-name"
          data-label="Mission Name"
        >
          <span
            class="data-table-cell-mission-name"
          >
            Hephaestus VI
          </span>
        </td>
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;

exports[`<DataTableCell /> with a column with value: function should match the snapshot 1`] = `
<DocumentFragment>
  <table>
    <tbody>
      <tr>
        <td
          class="data-table-cell-mission-name"
          data-label="Mission Name"
        >
          Hephaestus VI
        </td>
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;
